lang/src/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt

```diff
@@ -524,6 +525,93 @@ internal class StaticTypeInferenceVisitorTransform(
             return nAry.withStaticType(computeReturnTypeForFunctionCall(signature, functionArguments, nAry.metas))
         }
 
+        // Call agg : "count", "avg", "max", "min", "sum"
+        override fun transformExprCallAgg(node: PartiqlAst.Expr.CallAgg): PartiqlAst.Expr {
+            val nAry = super.transformExprCallAgg(node) as PartiqlAst.Expr.CallAgg
+            val funcName = nAry.funcName
+            // unwrap the type if this is a collectionType
+            val argType = when (val type = nAry.arg.getStaticType()) {
+                is CollectionType -> type.elementType
+                else -> type
+            }
+
+            return nAry.withStaticType(computeReturnTypeForAggFunc(funcName, argType))
+        }
+
+        private fun computeReturnTypeForAggFunc(aggFunc: SymbolPrimitive, argType: StaticType): StaticType {
+            // nested type is not supported by agg functions.
+            // i.e. sum(1,2,[3,4]) is not supported
+            val argTypeCheck: (StaticType) -> Boolean = {
+                it.isSubTypeOf(StaticType.unionOf(StaticType.MISSING, StaticType.NULL, StaticType.NUMERIC))
+            }
+            return when (aggFunc.text) {
+                // current implementation of count will always return a long
+                "count" -> StaticType.INT8
+                // max/min supports all type and the result depends on comparison.
+                "max", "min" -> argType
+                // current implementation of avg always return a decimal.
+                "avg" -> {
+                    if (argTypeCheck(argType)) StaticType.DECIMAL else error("Aggregate function AVG has incompatible input type")
```

Comment:
Don't think it always returns a decimal. E.g.:

```
PartiQL> AVG([1, 1]);
AVG([1, 1])==='
1
---
OK!
```

Perhaps it should return the union of all number types?

Reply:
https://github.com/partiql/partiql-lang-kotlin/blob/f96186e435f65c982372de17b6a05adbf049ab83/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L243-L260

Let me know if my understanding is wrong, but I think the current implementation for `AVG` indeed always return decimal(or `null`). The `1` in you example should be equivalent to `1d0`. 

Now, perhaps in your example the query should return an integer...

Reply:
Ah I see. The decimal is not showing up in the REPL for some reason. You're correct that the return type is numeric/decimal:

https://github.com/partiql/partiql-lang-kotlin/blob/f96186e435f65c982372de17b6a05adbf049ab83/lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt#L946-L951